pkg/qrcodegenerator/qrcode/gateways/ttigpro1.go

```diff
@@ -0,0 +1,140 @@
+// Copyright © 2024 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gateways
+
+import (
+	"regexp"
+	"strings"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/types"
+)
+
+const (
+	formatIDttigpro1 = "ttigpro1"
+
+	euiLength        = 16
+	OwnerTokenLength = 12
+)
+
+// TTIgpro1 is a format for gateway identification QR codes.
+type TTIgpro1 struct {
+	GtwEUI     types.EUI64 // lowercase base16, 16 chars long
+	OwnerToken []byte      // base62, 12 chars long
+}
+
+// ttigpro1Regex is the regular expression to match the TTIgpro1 format.
+// The format is as follows: https://ttig.pro/c/{16 lowercase base16 chars}/{12 base62 chars}
+var ttigpro1Regex = regexp.MustCompile(`^https://ttig\.pro/c/([a-f0-9]{16})/([a-zA-Z0-9]{12})$`)
+
+// Validate implements the Data interface.
+func (m TTIgpro1) Validate() error {
+	if err := m.validateOwnerToken(m.OwnerToken); err != nil {
+		return err
+	}
+	if err := m.validateGtwEUI(m.GtwEUI.String()); err != nil {
```

Comment:
Why is this even necessary? Unmarshaling already failed if the format is wrong right?

Reply:
I implemented the validation methods because the interface that the format must follow looks like this:
```go
// Data represents gateway QR code data.
type Data interface {
	qrcode.Data
	// FormatID returns the ID of the format used to parse the QR Code data.
	FormatID() string
}
```

Where the embedded qrcode.Data interface has the following definition:

```go
// Data represents QR code data.
type Data interface {
	Validate() error
	encoding.TextMarshaler
	encoding.TextUnmarshaler
}
```

To be consistent with the rest of the package, the Validate method (which also should be used in the TextMarshaler method too) has to be implemented. Otherwise, we have to define a separate interface for the gateway qrs.

Reply:
Eventually I dropped the embedded qrcode.Data interface, and kept only this:
```
// Data represents gateway QR code data.
type Data interface {
	// FormatID returns the ID of the format used to parse the QR Code data.
	FormatID() string
	GatewayEUI() string
	OwnerToken() string
	encoding.TextUnmarshaler
}
```

Let me know if this is okay.